### PR TITLE
feat: Add alignmentOffset to DropdownMenu

### LINKED ISSUE #151524: Add spacing between `MenuAnchor` and `dropdownMenuEntries` in `DropdownMenu`
### Use case

I am trying to build an interface like the attached image using `DropdownMenu` where there is space between `MenuAnchor` and `dropdownMenuEntries`.

Screenshot 2024-07-10 at 4 12 05 PM

### Proposal

Would you guys be open to accepting a pull request if i were to add an optional property called `spacing` to `DropdownMenu` and add a `SizedBox` to add a margin between `MenuAnchor` and `dropdownMenuEntries`? I think it would look something like below

```dart
return _DropdownMenuBody(
          width: widget.width,
          children: <Widget>[
            textField,
            SizedBox(height: spacing),
            for (final Widget item in _initialMenu!) item,
            trailingButton,
            leadingButton,
          ],
        );
```

### PR DESCRIPTION
**Pull request by jasonkang14**
Changes:
- Introduced `alignmentOffset` property to `DropdownMenu` to allow adjustment of menu alignment.
- Updated the `MenuAnchor` widget to utilize the new `alignmentOffset` property.
- Default value for `alignmentOffset` is `Offset.zero`, meaning no additional spacing by default.

Motivation: 
- This PR closes #151524 
- @nate-thegrate, please let me know if I need to make changes in order for this PR to be merged

The motivation behind this change is to provide developers with more control over the spacing between the `MenuAnchor` and `dropdownMenuEntries` in the `DropdownMenu` widget. This enhancement allows for better alignment and customization of the dropdown menu appearance.
| before | after |
| :---: | :---: |
| Screenshot 2024-07-14 at 8 03 14 PM      |      Screenshot 2024-07-14 at 8 03 27 PM   |

Initially, it was suggested to use a `SizedBox` to introduce spacing. However, upon further examination of the source code, it was discovered that the `DropdownMenuEntries` are rendered on the screen via an `OverlayPortal`. This necessitated leveraging the existing `alignmentOffset` property within the `MenuAnchor` for a more seamless and effective alignment adjustment.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

**Comment by jasonkang14** [comment]
I have signed the CLA but it appears that i have not. Adding a screenshot below as a proof

Screenshot 2024-07-14 at 8 17 58 PM

**Comment by jasonkang14** [comment]
it looks like the test which is not a part of this PR is failing

File: packages/flutter/lib/src/material/dropdown_menu.dart
Comment (nate-thegrate) [inline packages/flutter/lib/src/material/dropdown_menu.dart:479]: [`MenuAnchor.alignmentOffset`](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/material/menu_anchor.dart#L176-L195) already has detailed and descriptive documentation; it'd be nice to include that here as well.

We can use dartdoc templates to avoid having 2 copies of the same documentation:

```diff
+ /// {@template flutter.material.MenuAnchor.alignmentOffset}
  /// The offset of the menu relative to the alignment origin determined by
  /// [MenuStyle.alignment] on the [style] attribute and the ambient
  /// [Directionality].
  ///
  /// Use this for adjustments of the menu placement.
  ///
  /// Increasing [Offset.dy] values of [alignmentOffset] move the menu position
  /// down.
  ///
  /// If the [MenuStyle.alignment] from [style] is not an [AlignmentDirectional]
  /// (e.g. [Alignment]), then increasing [Offset.dx] values of
  /// [alignmentOffset] move the menu position to the right.
  ///
  /// If the [MenuStyle.alignment] from [style] is an [AlignmentDirectional],
  /// then in a [TextDirection.ltr] [Directionality], increasing [Offset.dx]
  /// values of [alignmentOffset] move the menu position to the right. In a
  /// [TextDirection.rtl] directionality, increasing [Offset.dx] values of
  /// [alignmentOffset] move the menu position to the left.
  ///
  /// Defaults to [Offset.zero].
+ /// {@endtemplate}
  final Offset? alignmentOffset;
```

And then here, we'd just use

```suggestion
  /// {@macro flutter.material.MenuAnchor.alignmentOffset}
```

Then both members would show the same explanation when you hover your mouse in VS Code, and when you look it up on [api.flutter.dev](https://api.flutter.dev/flutter/material/MenuAnchor/alignmentOffset.html).

File: packages/flutter/test/material/dropdown_menu_test.dart
Comment (nate-thegrate) [inline packages/flutter/test/material/dropdown_menu_test.dart:2436]: One of the [Flutter style guidelines](https://github.com/flutter/flutter/blob/82b63ff27d7a15581837c84ca9304d3023934545/docs/contributing/Style-guide-for-Flutter-repo.md#avoid-useless-documentation) is "avoid useless documentation". Adding comments can be helpful but also carries a risk: in the future someone could change the logic and forget to update the documentation.

In this test, I believe the code is "self-documenting" and doesn't require additional explanation: `find.byKey()` finds something by its key, and `expect(dropdownMenu.alignmentOffset, Offset.zero)` means that the test expects a value of zero for the alignment offset.

**Review by nate-thegrate — APPROVED** [review]
Phenomenal work here, leveraging an existing `MenuAnchor` parameter to fit the desired use case 👍

I have a couple of suggestions related to writing comments, but everything about the implementation looks fantastic!

**Comment by nate-thegrate** [comment]
In order to merge, we'll need 1 more approval, and all tests need to pass.

It looks like the test failure might be related to these changes:

```
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following TestFailure was thrown running a test:
Expected: null
  Actual: _SubmenuButtonState:<_SubmenuButtonState#8abcc(lifecycle state: defunct, not mounted)>

When the exception was thrown, this was the stack:
#4      main.<anonymous closure> (file:///Volumes/Work/s/w/ir/x/w/flutter/packages/flutter/test/material/menu_anchor_test.dart:3888:7)
<asynchronous suspension>
#5      testWidgets.<anonymous closure>.<anonymous closure> (package:flutter_test/src/widget_tester.dart:189:15)
<asynchronous suspension>
#6      TestWidgetsFlutterBinding._runTestBody (package:flutter_test/src/binding.dart:1032:5)
<asynchronous suspension>
<asynchronous suspension>
(elided one frame from package:stack_trace)

This was caught by the test expectation on the following line:
  file:///Volumes/Work/s/w/ir/x/w/flutter/packages/flutter/test/material/menu_anchor_test.dart line 3888
The test description was:
  Garbage collector destroys child _MenuAnchorState after parent is closed
════════════════════════════════════════════════════════════════════════════════════════════════════
```

File: packages/flutter/lib/src/material/dropdown_menu.dart
Comment (jasonkang14) [reply packages/flutter/lib/src/material/dropdown_menu.dart]: I never knew this. Thank you for this because I think this would benefit our project at work as well!

File: packages/flutter/test/material/dropdown_menu_test.dart
Comment (jasonkang14) [reply packages/flutter/test/material/dropdown_menu_test.dart:2436]: I removed the comments from the test code. I initially included them to clarify my intentions, but I realize now that the code is straightforward enough on its own.

Comment (nate-thegrate) [reply packages/flutter/test/material/dropdown_menu_test.dart:2436]: Perfect, thank you!

File: packages/flutter/lib/src/material/dropdown_menu.dart
Comment (nate-thegrate) [reply packages/flutter/lib/src/material/dropdown_menu.dart]: no prob :)

File: packages/flutter/lib/src/material/dropdown_menu.dart
Comment (nate-thegrate) [inline packages/flutter/lib/src/material/dropdown_menu.dart:175]: The dartdoc template looks great, and thanks for making the value here nullable to match!

Since a null value will default to `Offset.zero`, we can remove the explicit default argument here:

```suggestion
    this.alignmentOffset,
```

**Comment by jasonkang14** [comment]
I've implemented all the suggestions. Thank you. If the test code error originates elsewhere, is this PR ready for approval? The tests pass in my local environment, so if I need to make more changes, please guide me on how to reproduce the failed cases so I can make the necessary adjustments!

File: packages/flutter/lib/src/material/dropdown_menu.dart
Comment (jasonkang14) [reply packages/flutter/lib/src/material/dropdown_menu.dart]: Done! I appreciate your prompt reply.

**Comment by nate-thegrate** [comment]
> is this PR ready for approval?

Yes! Assuming all the checks pass, the only thing we need is a second reviewer. Some folks on the Flutter team are out on vacation, but I bet we'll be able to get approval number 2 sometime this week.

File: packages/flutter/test/material/dropdown_menu_test.dart
Comment (nate-thegrate) [inline packages/flutter/test/material/dropdown_menu_test.dart:2454]: It looks like defaulting to `null` caused this test to fail.

Perhaps the test could be updated as follows:
```dart
  testWidgets('DropdownMenu with an alignmentOffset does not throw errors', (WidgetTester tester) async {

    await tester.pumpWidget(
      MaterialApp(
        home: Scaffold(
          body: DropdownMenu<String>(
            alignmentOffset: Offset(1.0, 1.0),
            dropdownMenuEntries: const <DropdownMenuEntry<String>>[
              DropdownMenuEntry<String>(value: '1', label: 'One'),
              DropdownMenuEntry<String>(value: '2', label: 'Two'),
            ],
          ),
        ),
      ),
    );

    expect(tester.takeException(), isNull);
  });
```

**Comment by jasonkang14** [comment]
I think the test was failing because I got rid of the default `alignmentOffset` but did not update the test case accordingly. I just changed the test case as you suggested!

**Comment by jasonkang14** [comment]
While the exact error message from the [log](https://ci.chromium.org/ui/p/flutter/builders/try/Linux%20analyze/82753/overview) was not visible, running dart --enable-asserts analyze.dart locally revealed detailed error information. The following changes were made to resolve the issue:

**1. Updated Dependencies:**

- Added `flutter` SDK dependency in dev/snippets/pubspec.yaml.
- Ensured compatibility by updating the versions of various dependencies as shown below:

```yaml
dependencies:
  flutter:
    sdk: flutter
  analyzer: 6.4.1
  args: 2.4.2
  dart_style: 2.3.6
  file: 7.0.0
  meta: 1.12.0
  path: 1.9.0
  platform: 3.1.4
  process: 5.0.2
```

**2. Refactored StockStrings Classes:**

- Modified constructors to use the `super` keyword for better readability and maintainability.

**Before:**
```dart
// dev/benchmarks/test_apps/stocks/lib/i18n/stock_strings_en.dart

class StockStringsEn extends StockStrings {
  StockStringsEn([String locale = 'en']) : super(locale);
}
```

**After:**
```dart
// dev/benchmarks/test_apps/stocks/lib/i18n/stock_strings_en.dart

class StockStringsEn extends StockStrings {
  StockStringsEn([super.locale = 'en']);
}
```

- Applied similar changes to `StockStringsEs`:

**Before:**

```dart
// dev/benchmarks/test_apps/stocks/lib/i18n/stock_strings_es.dart

class StockStringsEs extends StockStrings {
  StockStringsEs([String locale = 'es']) : super(locale);
}
```


**After:**

```dart
// dev/benchmarks/test_apps/stocks/lib/i18n/stock_strings_es.dart

class StockStringsEs extends StockStrings {
  StockStringsEs([super.locale = 'es']);
}
```

- Refactored the `StockStrings` abstract class for consistency and clarity:

**Before:**

```dart
// dev/benchmarks/test_apps/stocks/lib/i18n/stock_strings.dart

abstract class StockStrings {
  StockStrings(String locale) : localeName = intl.Intl.canonicalizedLocale(locale.toString());
}
```

**After:**

```dart
// dev/benchmarks/test_apps/stocks/lib/i18n/stock_strings.dart

abstract class StockStrings {
  StockStrings(String locale) : localeName = intl.Intl.canonicalizedLocale(locale);
}
```

These modifications resolve the error, but I'm uncertain if they align with Flutter's design principles. I'd appreciate your review to determine if these changes are appropriate. If so, should we consider them for a separate pull request?

File: packages/flutter/test/material/dropdown_menu_test.dart
Comment (nate-thegrate) [inline packages/flutter/test/material/dropdown_menu_test.dart:2439]:
```suggestion
      const MaterialApp(
```

**Review by nate-thegrate — COMMENTED** [review]
Dang, sorry you had to go through all that!

In the future, you can see the full log by going to the [overview page](https://ci.chromium.org/ui/p/flutter/builders/try/Linux%20analyze/82753/overview), and then scrolling down & clicking on either "stdout" or "[raw]":

image

Looks like the problem can be fixed with a `const` constructor.

**Comment by jasonkang14** [comment]
@nate-thegrate finally passed all the tests. thank you!

File: packages/flutter/test/material/dropdown_menu_test.dart
Comment (navaronbracke) [inline packages/flutter/test/material/dropdown_menu_test.dart:2436]: I do not think that this test is testing the right thing? The `MenuAnchor` widget accepts a nullable alignmentOffset. So it would not "crash" if you gave it an alignment offset.

Instead, grab the `MenuAnchor` from the subtree and compare the alignmentOffset in an expect:

```
final MenuAnchor menuAnchor = tester.widget<MenuAnchor>(find.byType(MenuAnchor));

expect(menuAnchor.alignmentOffset, alignmentOffset);
```

**Comment by jasonkang14** [comment]
I have implemented the suggested changes to the test case provided by @navaronbracke . and, I have rebased this branch on the latest master as well

File: packages/flutter/test/material/dropdown_menu_test.dart
Comment (nate-thegrate) [reply packages/flutter/test/material/dropdown_menu_test.dart]: To expand on this a bit: This PR adds the `alignmentOffset` parameter to `DropdownMenu`, when previously it could only be configured for `MenuAnchor`. So if the changes to dropdown_menu.dart were reverted, attempting to set an offset this way would throw an exception.

With that in mind, I think the new version of this test is much better than what I'd suggested beforehand 👍

**Comment by jasonkang14** [comment]
I have rebased this branch on the latest master because `tree-status` check was failing but now there is an issue with `Google testing`. and the log just says "Please contact a Google developer to investigate." do you guys have any suggestions for me to pass this test?

**Comment by nate-thegrate** [comment]
If the **tree-status** is failing, it just means that there are testing issues on the main branch. It usually switches back and forth a few times each day; if it's failing, we can add the `autosubmit` label, and a bot will automatically merge the PR once it's back to green.

**Google testing** failures could be a number of different things. Unfortunately, not everyone with [commit access](https://github.com/flutter/flutter/blob/master/docs/contributing/Contributor-access.md) can see the test output, so we'll need to wait for a Googler to take a look.

**Comment by jasonkang14** [comment]
@nate-thegrate appreciate the prompt response! thank you :)

**Review by QuncCccccc — APPROVED** [review]
LGTM! Thanks for the contribution:)

**Comment by SergeyShustikov** [comment]
Hi, how soon it will be released?

**Comment by nate-thegrate** [comment]
> Hi, how soon it will be in stable version?

This fix is available in the current beta channel and will be included in the next stable release (scheduled for mid-November).

**Pull request merged**
